File: breakout.rb
#!/usr/bin/env ruby
# breakout.rb
# encoding: UTF-8

require 'curses'
require 'json'
require 'fileutils'

RECORD_FILE = File.join(Dir.home, '.breakout_record.json')
def load_record
  return 0 unless File.exist?(RECORD_FILE)
  JSON.parse(File.read(RECORD_FILE))['record'] || 0
rescue
  0
end
def save_record(record)
  File.write(RECORD_FILE, JSON.pretty_generate(record: record))
end

Curses.init_screen
Curses.start_color
Curses.use_default_colors
(1..7).each { |i| Curses.init_pair(i, i, -1) }

height = Curses.lines
width = Curses.cols
if height < 25 || width < 50
  puts "Terminal too small"
  exit 1
end

speed = 50
if ARGV.include?('-s') && ARGV.index('-s') + 1 < ARGV.size
  speed = ARGV[ARGV.index('-s') + 1].to_i
end

H = height - 2
W = width - 2
padY = 1; padX = 1

paddleW = 8
paddleX = (W - paddleW) / 2
paddleY = H - 2
ballX = paddleX + paddleW/2.0
ballY = paddleY - 1
ballDX = 1.0
ballDY = -1.0
ballSpeed = 1.5
ballActive = false
gameOver = false
lives = 3
score = 0
level = 1
best = load_record
frame = speed / 1000.0

blocks = []
bonuses = []

def generate_level(blocks, level, W)
  blocks.clear
  rows = 5 + level/2
  cols = 8 + level
  rows.times do |r|
    cols.times do |c|
      if c*2+2 < W-2
        hp = rand(100) < 70 ? 1 : (rand(100) < 50 ? 2 : 3)
        color = [hp+1, 7].min
        blocks << {x: c*2+2, y: r*2+2, hp: hp, color: color}
      end
    end
  end
end
generate_level(blocks, level, W)

Curses.curs_set(0)
Curses.noecho
Curses.timeout=0

loop do
  ch = Curses.getch
  if ch == 'q' || ch == 'Q'
    break
  elsif ch == 'r' || ch == 'R'
    if gameOver
      paddleX = (W-paddleW)/2
      ballX = paddleX + paddleW/2.0
      ballY = paddleY - 1
      ballDX = 1.0
      ballDY = -1.0
      ballActive = false
      lives = 3
      score = 0
      level = 1
      gameOver = false
      bonuses.clear
      paddleW = 8
      ballSpeed = 1.5
      generate_level(blocks, level, W)
      next
    end
  elsif ch == ' '
    if !ballActive && !gameOver
      ballActive = true
    end
  elsif ch == Curses::KEY_LEFT || ch == 'a'
    paddleX = [0, paddleX - 2].max
  elsif ch == Curses::KEY_RIGHT || ch == 'd'
    paddleX = [W - paddleW, paddleX + 2].min
  end

  if gameOver
    Curses.clear
    msg = "GAME OVER! Score: #{score}  Best: #{best}"
    Curses.setpos(H/2-2, (W - msg.length)/2)
    Curses.attron(Curses.color_pair(4)) { Curses.addstr(msg) }
    Curses.setpos(H/2, (W-20)/2)
    Curses.attron(Curses.color_pair(4)) { Curses.addstr("R - restart | Q - quit") }
    Curses.refresh
    next
  end

  if ballActive
    ballX += ballDX * ballSpeed
    ballY += ballDY * ballSpeed
    if ballX <= 0 || ballX >= W-1
      ballDX *= -1
    end
    if ballY <= 0
      ballDY *= -1
    end
    if ballY.to_i == paddleY-1 && paddleX <= ballX.to_i && ballX.to_i < paddleX+paddleW
      ballDY *= -1
      offset = (ballX - paddleX) / paddleW
      ballDX = (offset - 0.5) * 2
      if ballDX.abs < 0.3
        ballDX = ballDX >= 0 ? 0.5 : -0.5
      end
      print "\a"
    end
    blocks.each do |b|
      if b[:x] <= ballX && ballX < b[:x]+2 && b[:y] <= ballY && ballY < b[:y]+2
        b[:hp] -= 1
        if b[:hp] <= 0
          score += 10
          if rand(100) < 15
            type = rand(4)
            bonuses << {x: b[:x], y: b[:y], type: type}
          end
          blocks.delete(b)
          print "\a"
        end
        ballDY *= -1
        break
      end
    end
    if ballY >= H
      lives -= 1
      if lives <= 0
        gameOver = true
        if score > best
          best = score
          save_record(best)
        end
      else
        ballX = paddleX + paddleW/2.0
        ballY = paddleY - 1
        ballDX = 1.0
        ballDY = -1.0
        ballActive = false
      end
    end
    if blocks.empty?
      level += 1
      ballSpeed += 0.3
      generate_level(blocks, level, W)
      ballX = paddleX + paddleW/2.0
      ballY = paddleY - 1
      ballDX = 1.0
      ballDY = -1.0
      ballActive = false
    end
  end

  bonuses.each do |b|
    b[:y] += 1
    if b[:y] >= H
      bonuses.delete(b)
      next
    end
    if b[:y] == paddleY && paddleX <= b[:x] && b[:x] < paddleX+paddleW
      case b[:type]
      when 0 then paddleW = [16, paddleW+4].min
      when 1 then ballSpeed += 0.5
      when 2 then lives += 1
      when 3
        blocks.each { |bb| bb[:hp] -= 1 }
      end
      bonuses.delete(b)
      print "\a"
    end
  end

  # Draw
  Curses.clear
  (0..H).each do |y|
    Curses.setpos(padY+y, padX-1); Curses.addstr('|')
    Curses.setpos(padY+y, padX+W); Curses.addstr('|')
  end
  (0..W+1).each do |x|
    Curses.setpos(padY-1, padX+x-1); Curses.addstr('-')
  end
  (0...paddleW).each do |i|
    Curses.setpos(padY+paddleY, padX+paddleX+i)
    Curses.attron(Curses.color_pair(1)) { Curses.addstr('=') }
  end
  if ballActive || !gameOver
    Curses.setpos(padY+ballY.to_i, padX+ballX.to_i)
    Curses.attron(Curses.color_pair(2)) { Curses.addstr('O') }
  end
  blocks.each do |b|
    Curses.attron(Curses.color_pair(b[:color]))
    (0...2).each do |i|
      (0...2).each do |j|
        Curses.setpos(padY+b[:y]+j, padX+b[:x]+i)
        Curses.addstr('#')
      end
    end
    Curses.attroff(Curses.color_pair(b[:color]))
  end
  bonuses.each do |b|
    sym = ['W','S','L','F'][b[:type]]
    Curses.setpos(padY+b[:y], padX+b[:x])
    Curses.attron(Curses.color_pair(3)) { Curses.addstr(sym) }
  end
  Curses.attron(Curses.color_pair(4))
  Curses.setpos(0, 2); Curses.addstr("Score: #{score}")
  Curses.setpos(0, W/2-4); Curses.addstr("Best: #{best}")
  Curses.setpos(0, W-20); Curses.addstr("Lives: #{lives}  Level: #{level}")
  Curses.attroff(Curses.color_pair(4))
  if gameOver
    msg = "GAME OVER! Press R to restart, Q to quit"
    Curses.setpos(H/2, (W-msg.length)/2)
    Curses.attron(Curses.color_pair(5)) { Curses.addstr(msg) }
  end
  Curses.refresh
  sleep(frame)
end

Curses.close_screen

File: breakout.cpp
// breakout.cpp
#include <curses.h>
#include <stdlib.h>
#include <time.h>
#include <unistd.h>
#include <fstream>
#include <string>
#include <vector>
#include <json/json.h>

using namespace std;

int loadRecord() {
    ifstream f(getenv("HOME") + string("/.breakout_record.json"));
    Json::Value root;
    if (f >> root) return root["record"].asInt();
    return 0;
}

void saveRecord(int record) {
    Json::Value root;
    root["record"] = record;
    ofstream f(getenv("HOME") + string("/.breakout_record.json"));
    f << root.toStyledString();
}

struct Block { int x, y, hp, color; };
struct Bonus { int x, y, type; };

int main(int argc, char* argv[]) {
    int speed = 50;
    for (int i=1; i<argc; ++i) {
        if (string(argv[i]) == "-s" && i+1 < argc) speed = atoi(argv[++i]);
        else if (string(argv[i]) == "-h") { cout << "Usage: breakout [-s speed_ms]\n"; return 0; }
    }

    initscr();
    cbreak();
    noecho();
    curs_set(0);
    nodelay(stdscr, TRUE);
    keypad(stdscr, TRUE);
    start_color();
    for (int i=1; i<=7; ++i) init_pair(i, i, COLOR_BLACK);

    int height, width;
    getmaxyx(stdscr, height, width);
    if (height < 25 || width < 50) { endwin(); cout << "Terminal too small\n"; return 1; }

    int H = height-2, W = width-2;
    int pad_y=1, pad_x=1;

    int paddle_w=8, paddle_x=(W-paddle_w)/2, paddle_y=H-2;
    float ball_x=paddle_x+paddle_w/2, ball_y=paddle_y-1;
    float ball_dx=1, ball_dy=-1, ball_speed=1.5;
    bool ball_active=false, game_over=false;
    int lives=3, score=0, level=1;
    int best=loadRecord();
    int frame = speed*1000;

    vector<Block> blocks;
    vector<Bonus> bonuses;

    auto generate_level = [&]() {
        blocks.clear();
        int rows = 5 + level/2;
        int cols = 8 + level;
        for (int r=0; r<rows; ++r) {
            for (int c=0; c<cols; ++c) {
                if (c*2+2 < W-2) {
                    int hp = (rand()%100 < 70) ? 1 : (rand()%100 < 50 ? 2 : 3);
                    int color = min(hp+1, 7);
                    blocks.push_back({c*2+2, r*2+2, hp, color});
                }
            }
        }
    };
    generate_level();

    auto draw = [&]() {
        clear();
        for (int y=0; y<=H; ++y) { mvaddch(pad_y+y, pad_x-1, '|'); mvaddch(pad_y+y, pad_x+W, '|'); }
        for (int x=0; x<W+2; ++x) mvaddch(pad_y-1, pad_x+x-1, '-');
        attron(COLOR_PAIR(1));
        for (int i=0; i<paddle_w; ++i) mvaddch(pad_y+paddle_y, pad_x+paddle_x+i, '=');
        attroff(COLOR_PAIR(1));
        if (ball_active || !game_over) {
            attron(COLOR_PAIR(2));
            mvaddch(pad_y+(int)ball_y, pad_x+(int)ball_x, 'O');
            attroff(COLOR_PAIR(2));
        }
        for (auto &b : blocks) {
            attron(COLOR_PAIR(b.color));
            for (int i=0; i<2; ++i) for (int j=0; j<2; ++j)
                mvaddch(pad_y+b.y+j, pad_x+b.x+i, '#');
            attroff(COLOR_PAIR(b.color));
        }
        attron(COLOR_PAIR(3));
        for (auto &bon : bonuses) {
            char sym = (bon.type==0)?'W':(bon.type==1)?'S':(bon.type==2)?'L':'F';
            mvaddch(pad_y+bon.y, pad_x+bon.x, sym);
        }
        attroff(COLOR_PAIR(3));
        attron(COLOR_PAIR(4));
        mvprintw(0, 2, "Score: %d", score);
        mvprintw(0, W/2-4, "Best: %d", best);
        mvprintw(0, W-20, "Lives: %d  Level: %d", lives, level);
        attroff(COLOR_PAIR(4));
        if (game_over) {
            const char* msg = "GAME OVER! Press R to restart, Q to quit";
            mvprintw(H/2, (W - strlen(msg))/2, "%s", msg);
        }
        refresh();
    };

    while (true) {
        int ch = getch();
        if (ch=='q'||ch=='Q') break;
        if (ch=='r'||ch=='R') {
            if (game_over) {
                paddle_x=(W-paddle_w)/2; ball_x=paddle_x+paddle_w/2; ball_y=paddle_y-1;
                ball_dx=1; ball_dy=-1; ball_active=false;
                lives=3; score=0; level=1; game_over=false;
                bonuses.clear(); paddle_w=8; ball_speed=1.5;
                generate_level();
                continue;
            }
        }
        if (ch==' ') {
            if (!ball_active && !game_over) ball_active = true;
        }

        if (game_over) { draw(); continue; }

        if (ch==KEY_LEFT || ch=='a') paddle_x = max(0, paddle_x-2);
        else if (ch==KEY_RIGHT || ch=='d') paddle_x = min(W-paddle_w, paddle_x+2);

        if (ball_active) {
            ball_x += ball_dx * ball_speed;
            ball_y += ball_dy * ball_speed;
            if (ball_x <= 0 || ball_x >= W-1) ball_dx *= -1;
            if (ball_y <= 0) ball_dy *= -1;
            if ((int)ball_y == paddle_y - 1 && paddle_x <= (int)ball_x && (int)ball_x < paddle_x + paddle_w) {
                ball_dy *= -1;
                float offset = (ball_x - paddle_x) / paddle_w;
                ball_dx = (offset - 0.5) * 2;
                if (abs(ball_dx) < 0.3) ball_dx = (ball_dx >= 0) ? 0.5 : -0.5;
                putchar('\a');
            }
            for (auto it=blocks.begin(); it!=blocks.end(); ++it) {
                if (it->x <= ball_x && ball_x < it->x+2 &&
                    it->y <= ball_y && ball_y < it->y+2) {
                    it->hp--;
                    if (it->hp <= 0) {
                        score += 10;
                        if (rand()%100 < 15) {
                            int btype = rand()%4;
                            bonuses.push_back({it->x, it->y, btype});
                        }
                        blocks.erase(it);
                        putchar('\a');
                    }
                    ball_dy *= -1;
                    break;
                }
            }
            if (ball_y >= H) {
                lives--;
                if (lives <= 0) { game_over=true; if(score>best){best=score; saveRecord(best);} }
                else { ball_x=paddle_x+paddle_w/2; ball_y=paddle_y-1; ball_dx=1; ball_dy=-1; ball_active=false; }
            }
            if (blocks.empty()) {
                level++; ball_speed += 0.3;
                generate_level();
                ball_x=paddle_x+paddle_w/2; ball_y=paddle_y-1;
                ball_dx=1; ball_dy=-1; ball_active=false;
            }
        }

        for (auto it=bonuses.begin(); it!=bonuses.end(); ) {
            it->y += 1;
            if (it->y >= H) { it = bonuses.erase(it); continue; }
            if (it->y == paddle_y && paddle_x <= it->x && it->x < paddle_x + paddle_w) {
                if (it->type == 0) paddle_w = min(16, paddle_w+4);
                else if (it->type == 1) ball_speed += 0.5;
                else if (it->type == 2) lives++;
                else if (it->type == 3) {
                    for (auto &b : blocks) b.hp -= 1;
                }
                it = bonuses.erase(it);
                putchar('\a');
            } else ++it;
        }

        draw();
        usleep(frame);
    }
    endwin();
    return 0;
}
